cohere/finetune_dataset.py

```diff
@@ -0,0 +1,171 @@
+
+    @abstractmethod
+    def get_train_data(self) -> Iterable[bytes]:
+        ...
+
+    @abstractmethod
+    def get_eval_data(self) -> Iterable[bytes]:
+        ...
+
+
+class LocalFileDataset(Dataset):
+    def __init__(self, train_file: str, eval_file: Optional[str] = None):
+        self._train_file = train_file
+        self._eval_file = eval_file
+
+    @abstractmethod
+    def file_config(self) -> FileConfig:
+        pass
+
+    def train_file_name(self) -> str:
+        return Path(self._train_file).name
+
+    def eval_file_name(self) -> str:
+        if not self.has_eval_file():
+            raise ValueError("Dataset has no eval file")
+        return Path(self._eval_file).name
+
+    def has_eval_file(self) -> str:
+        return self._eval_file is not None
+
+    def get_train_data(self) -> Iterable[bytes]:
+        with open(self._train_file, mode="rb") as _f:
+            for line in _f:
+                yield line
+
+    def get_eval_data(self) -> Iterable[bytes]:
+        if not self.has_eval_file():
+            raise ValueError("Dataset has no eval file")
+        with open(self._eval_file, mode="rb") as _f:
+            for line in _f:
+                yield line
+
+
+class CsvDataset(LocalFileDataset):
+    """
+    A dataset consisting of local csv files. Each row should contain two items.
+    E.g.: for prompt completion:
+    this is the prompt,and this the completion
+    another prompt, another completion
+    """
+
+    def __init__(self, train_file: str, delimiter: str, eval_file: Optional[Path] = None):
+        """
+        Args:
+            train_file (str): local path to csv with training data
+            eval_file: (str, optional): local path to a csv with eval data
+        """
+        super().__init__(train_file, eval_file)
+        self._delimiter = delimiter
+
+    def file_config(self) -> FileConfig:
+        config = dict(_empty_file_config)
+        config["delimiter"] = self._delimiter
+        return config
+
+
+class JsonlDataset(LocalFileDataset):
+    """
+    A dataset consisting of local jsonl files.
+    Examples:
+        prompt completion: {"prompt": "this is the prompt", "completion": "this is the completion"}
+    """
+
+    def file_config(self) -> FileConfig:
+        return _empty_file_config.copy()
+
+
+class TextDataset(LocalFileDataset):
+    """
+    A dataset consisting of local text files. Can be used for generative finetunes.
+    """
+
+    def __init__(self, train_file: str, separator: Optional[str], eval_file: Optional[Path] = None):
+        super().__init__(train_file, eval_file)
+        self._separator = separator
+
+    def file_config(self) -> FileConfig:
+        config = _empty_file_config.copy()
+        if self._separator:
+            config["separator"] = self._separator
+        return config
+
+
+class InMemoryDataset(Dataset):
+    """
+    A dataset existing in memory. You may pass a generator to avoid loading your whole dataset into memory at once.
```

Comment:
suggests you can do this with less RAM than your dataset, but you really can't as the StringIO will store it anyway?

Reply:
yes but it is only for one row at a time. Or did I misunderstand how generators work?
```
for row in self._training_data:
             yield self._serialize_row(row)             
```

And then
```
def _serialize_row(self, row: tuple[str, str]) -> bytes:
         buffer = StringIO()
         writer = csv.writer(buffer, delimiter=self._delimiter)
         writer.writerow(row)
         return f"{buffer}\n".encode("utf-8")
```

Reply:
Sorry, you are right!